### PR TITLE
feat: [IC-1680] Add registry data provider for canisters

### PR DESCRIPTION
**Pull request by pietrodimarco-dfinity**
This registry data provider is designed to work with the `ic-registry-canister-client`
in canisters, enabling the retrieval and storage of a registry copy in stable memory.

* If `keys_to_keep` is `Some(keys)`, only the specified `keys` will be stored in stable memory,
  while all other keys from the registry will be discarded.
* If `keys_to_keep` is `None`, all keys from the registry will be retained in stable memory.

### CODE REVIEW

File: rs/registry/canister_data_provider/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        Ok(registry_delta)
+    }
+
+    fn get_latest_version(&self) -> Option<u64> {
+        S::with_borrow(|local_registry| local_registry.last_key_value().map(|(k, _)| k.version))
+    }
+
+    fn add_deltas(&self, deltas: Vec<RegistryDelta>) -> anyhow::Result<()> {
+        for delta in deltas {
+            let string_key = std::str::from_utf8(&delta.key[..])
```
Comment (sasa-tomic) [inline rs/registry/canister_data_provider/src/lib.rs:121]: We could switch this to `String::from_utf8_lossy` if the NNS team says that we may have non-utf8 keys?

Comment (pietrodimarco-dfinity) [reply rs/registry/canister_data_provider/src/lib.rs:121]: This is also used by the `nns_data_provider` but sure we could change both

Comment (max-dfinity) [reply rs/registry/canister_data_provider/src/lib.rs:121]: I don't think we have non utf8 keys

File: rs/registry/canister_data_provider/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    const BOUND: Bound = Bound::Unbounded;
+}
+
+#[derive(Clone, Ord, PartialOrd, Eq, PartialEq, Default)]
+pub struct StorableRegistryKey {
+    version: u64,
+    key: String,
+}
+
+// This value is set as 2 times the max key size present in the registry
```
Comment (sasa-tomic) [inline rs/registry/canister_data_provider/src/lib.rs:41]: Is the max key size limit == 100? or was this just based on observation of current data?

File: rs/registry/canister_data_provider/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+const MAX_REGISTRY_KEY_SIZE: u32 = 200;
```
Comment (sasa-tomic) [inline rs/registry/canister_data_provider/src/lib.rs:42]: What if in the future there are longer keys?

File: rs/registry/canister_data_provider/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "stable_data_provider"
```
Comment (sasa-tomic) [inline rs/registry/canister_data_provider/Cargo.toml:2]:
```suggestion
name = "ic-registry-canister-data-provider"
```
?

File: rs/registry/canister_data_provider/src/lib.rs
Comment (sasa-tomic) [inline rs/registry/canister_data_provider/src/lib.rs:1]: A tiny usage example (like main.rs or similar) might be a nice addition - to demonstrate how to use this code.

**Review by sasa-tomic — APPROVED** [review]
LGTM!

File: rs/registry/canister_data_provider/src/lib.rs (hunk 2 above)
Comment (pietrodimarco-dfinity) [reply rs/registry/canister_data_provider/src/lib.rs:41]: This is just the max key of the current data

File: rs/registry/canister_data_provider/src/lib.rs (hunk 3 above)
Comment (pietrodimarco-dfinity) [reply rs/registry/canister_data_provider/src/lib.rs:42]: We can always increase this number in the future

**Review by max-dfinity — APPROVED** [review]

File: rs/registry/canister_data_provider/src/lib.rs
Comment (pietrodimarco-dfinity) [reply rs/registry/canister_data_provider/src/lib.rs:1]: Good idea will add it ;)

**Pull request merged**